Makes the scalaTag (_?) configurable.

borissmidt opened the pull request:
issue: #80
Fix Archive appendix problem
Found undocumented weird behavior

borissmidt committed:
#80
Fix Archive appendix problem
Found undocumented weird behavior

borissmidt commented on src/main/groovy/com/github/prokod/gradle/crossbuild/utils/CrossBuildPluginUtils.groovy:
```diff
@@ -63,7 +63,13 @@ class CrossBuildPluginUtils {
     static String qmarkReplace(String template, String replacement) {
-        template.replaceAll('\\?', replacement)
+        String newTag = template.replaceAll('\\?', replacement)
+        if(newTag == template) {
```
I think this _? replacement should be either documented or removed.
imageine if some wnated to have the extsions _?.?? :roll_eyes:

borissmidt commented on src/test/groovy/com/github/prokod/gradle/crossbuild/BuildResolverSpec.groovy:
```diff
@@ -0,0 +1,23 @@
+
+        resolver.generateCrossArchivesNameAppndix("_2.11","2.12") == "_2.12"
+
+        resolver.generateCrossArchivesNameAppndix("_?_weird","2.12") == "_2.12_weird"
```
To me this seems like a very weird behavior of the plugin.

borissmidt commented:
By the way you already get the default scala version from the scala library anyway so why have _? as the postfix instead of the tagVersion detected from the scala lib? like if you have 2.12.11 as a dependency then  `_2.12` is the crossbuild tag? Then you don't have to enable this plugin on all sub projects like you have to do now. Otherwise it cannot resolve the _? depedencies.

prokod commented:
@borissmidt hi :)
I think you are touching couple of things here:
First this specific method is used only for the plugin DSL part and not for the dependencies under dependencies block.

Second, regarding the cooperation between different plugins like play plugin - this can be improved if you will send me a dummy full project I can look on.

I hope this makes sense, The test is misleading. I will update it so it won't be such misleading.

Noam

borissmidt commented:
Actually i tried to set the apendix to _2.11 so all other project would just see the dependency as the _2.11 depdency and be able to resolve it. But once i tried it then the dependencies in 2.12 got the appendix 2.11 as the dsl states. so then it complains that the version isn't correct. 


Hence this unit test and the fix. I will make a SNAPSHOT version locally to verify if this fix works and update the integration test of this project once i have time.

borissmidt commented:
Oke you are right it doesn't work and now the error message still gives the wrong dependency back.
I'll make a small example project but in general it is just the issue that you have to enable the plugin everywhere otherwise it cannot resolve the `_?` dependencies. 

I think it would be better to just be able to declare the appendix to `_2.11` so projects depending on each other would just get the right version if no crossbuild is configured. But this doesn't work and it complains that it cannot find a 2.12 version of lib `commons_2.11` for the `2.12` build. But it should have looked to `commons_2.12`.
Thus setting a custom appendix to `_2.11` doesn't work.

So we could make this plugin accept having a `_2.11` appendix it would simplify project setup that it doesn't need the crossbuild plugin on all projects.  Or better a flag to automaticaly set the appendix to the detected scala version. i.e. if you include scala 2.11.12 it will be _2.11 if you include 2.13.3 then it will be _2.13

the old play plugin is a bit more special. it has its own dependency definition. Thus even when i enable the plugin it will not be happy anyway.

borissmidt commented:
Oke the code is clear now the appendix is not the thing i expected it to be. So i think my feature request is to get a custom `scalaTag` indication. so i could override the _? to _2.11 or _% or anything else the user wants.

borissmidt committed:
#80
Added a setting to allow for a 'scalaTag' so you can set something else than _? to mark scala dependencies.

borissmidt committed:
#80
removed some prints.

borissmidt commented:
@prokod I've updated the PR so the scala tag (_?) is configurable now.

prokod commented:
Hi @borissmidt :)

Big work man, really appreciated!

Before we go and incorporate that in some form, I will really appreciate the example we have talked about. I have intended that the plugin "default" Scala version will be derived from the non crossbuild configuration.

So basically having
```
compile common_2.11
crossBuildV210Compile common_2.10
crossBuildV211Compile common_2.11
```

Should set default non cross compile to 2.11 similar to `scalaTag = _2.11`

Of course, if all dependencies are in the form of `compile common_?` There is no way to know the "default" Scala version. For that I have incorporated inference through `compile org.scala-lang:scala-library:2.11.x`. I am mentioning that in the readme, though not explicit enough, and maybe not specifically for this use case you are dealing with.

It is pretty clear there is some bug, if this is not what you are getting. I would like to try and follow my original intent prior to extending plugin DSL, and an example project will assist greatly.

Cheers man!

borissmidt commented:
Yes, i added a test for that. Your plugin only works if all projects have
the plugin enabled. So if you have project A with a dependency common_?
With your plugin enabled and then a project B depending on project A
without the plugin enabled it will tell me that it cannot resolve common_?.

On Tue, Jun 16, 2020, 22:33 Noam Asor <notifications@github.com> wrote:

> Hi @borissmidt <https://github.com/borissmidt> :)
>
> Big work man, really appreciated!
>
> Before we go and incorporate that in some form, I will really appreciate
> the example we have talked about. I have intended that the plugin "default"
> Scala version will be derived from the non crossbuild configuration.
>
> So basically having
>
> compile common_2.11
> crossBuildV210Compile common_2.10
> crossBuildV211Compile common_2.11
>
> Should set default non cross compile to 2.11 similar to scalaTag = _2.11
>
> It is pretty clear there is some bug, if this is not what you are getting.
> I would like to try and follow my original intent prior to extending plugin
> DSL.
>
> Cheers man!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/prokod/gradle-crossbuild-scala/pull/81#issuecomment-644996930>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABNXZFWDZWZK6SWXVWOLI43RW7JJ7ANCNFSM4N47APYQ>
> .
>

prokod commented:
Couple of ways to tackle that in a different way ? I am not saying those are better or worse but at least let's put those on the table so we can look at the current issue at hand and how to solve it in the correct approach.

1. To decouple project A and project B to two different repos (doesn't need any code change :) ) ?
2. To sense play plugin (old/new) and allow both plugins (play and crossbuild to coexist) ?

We can put pros and cons to each including the one that says as you have implemented, if there is no valid resolution to _? default to scalaTag

prokod commented:
BTW @borissmidt the PR is huge and it will save me time if you could direct me to the test that shows that the above is now working - projectA **with** crossBuild plugin and projectB **w/o** crossBuild plugin which depends on projectA

borissmidt commented:
The first one is something we actively avoid in our company we try to have
the api. For example a swagger or grpc api whihc is generated inside the
play project.

The second one of upgrading the play plugin to play.framework plugin. Is
something i tried as wel but this broke the project.

A tird option that i did wad to manually set the dependencies with the
specialized applied. So i set default compile commom_2.11 and then set
crossbuildV212 to common_2.12. but then i'm essentially doing something
the plugin should do. 

The added advantage by setting the scala tag to the lib versions is that
normal compileScala dependencies don't need any processing. The plugin just
replaces 2.11 with other versions and you can still override for
dependencies that don't have a 2.11 and 2.13 in the same version.

For example i would only enable the corssbuild plugin for things i
publish. All other sub projects should not have crossbuild enabled.

so in general i think this would even simplify the plugin. Since then you just apply it to any project and no need to change any of you dependencies to a ? marked dependency. you just set scalaTag to 2.11 or 2.12 or 2.13 and done it cross compiles that sub project for you.

borissmidt commented:
Yes the PR is usage sadly, maybe i should move the setting from archive to somewhere else so you can only configure it globally. But it seems to me that there is a small structure problem in the code that it needs to ripple trough so many places to implement this. Maybe i should have passed it by constructor instead.

borissmidt commented:
Is there some way i could improve this PR or simplify it by quite a bit.
i.e. could i get the extension from some 'global' variable to get the scala tag out of it so i don't have to ripple it through the whole project?

prokod commented:
I am thinking on it. I have an idea, it will be probably easier for me to try it out on a branch based on yours

borissmidt commented:
Cool, thank you for the help!

On Thu, Jun 18, 2020, 13:45 Noam Asor <notifications@github.com> wrote:

> I am thinking of it. I have an idea, it will be probably easier for me to
> try it out on a branch based on yours
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/prokod/gradle-crossbuild-scala/pull/81#issuecomment-645963895>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABNXZFV6GBIHICUOYDRVFSDRXH46ZANCNFSM4N47APYQ>
> .
>

borissmidt commented:
let me know if i can assist in any way :+1:

borissmidt commented:
Maybe its just much easier to put it into a  system property?

prokod commented:
@borissmidt Hi there :)

I have been reflecting on the subject and here are some things I resurfaced as reasoning for some of the past decisions
- The name of the plugin was not specifically named ...scala.. as I have used the plugin in a muti dimensional crosscompiling env. I am working a lot with Spark and when developing Spark libs quickly you find yourself maintaining not only multiple scala version but also multiple Spark versions.
- Related to first topic, this is why your approach as is won't be fully satisfying, at least from DSL point of view.

Last but not least! I have found a different way to tackle that problem, implemented it and ran the tests including the one you added where one of the sub projects does not have crossbuilding plugin applied to it and it passed. No need to add nothing to DSL. It is of course inspired from the need you have expressed and and the concept you proposed !
Basically I have added a stage where I auto convert `compile lib_?` to a set of explicit dependencies that can be resolved by any non crossbuild sub project:
```
crossBuildV211Compile lib_2.11
crossBuildV212Compile lib_2.12
compile lib_<default_scala_version>
```
where default_scala_version is inferred from the non crossbuild scala lib dep version
I will beautify it a bit and prepare new version based on it.

I hope this approach will cover your use case nicely.

borissmidt commented:
Yes sure thats good as wel! I think its a big boon for this if you have
apps that define there api in the same project as the app. Since the app
never has to be cross compiled.

On Sun, Jun 21, 2020, 19:43 Noam Asor <notifications@github.com> wrote:

> @borissmidt <https://github.com/borissmidt> Hi there :)
>
> I have been reflecting on the subject and here are some things I
> resurfaced as reasoning for some of the past decisions
>
>    - The name of the plugin was not specifically named ...scala.. as I
>    have used the plugin in a muti dimensional crosscompiling env. I am working
>    a lot with Spark and when developing Spark libs quickly you find yourself
>    maintaining not only multiple scala version but also multiple Spark
>    versions.
>    - Related to first topic, this is why your approach as is won't be
>    fully satisfying, at least from DSL point of view.
>
> Last but not least! I have found a different way to tackle that problem,
> implemented it and ran the tests including the one you added where one of
> the sub projects does not have crossbuilding plugin applied to it and it
> passed. No need to add nothing to DSL. It is of course inspired from the
> need you have expressed and I believe a lot will benefit from !
> Basically I have added a stage where I auto convert compile lib_? to a
> set of explicit dependencies that can be resolved by any non crossbuild sub
> project:
>
> crossBuildV211Compile lib_2.11
> crossBuildV212Compile lib_2.12
> compile lib_<default_scala_version>
>
> where default_scala_version is inferred from the non crossbuild scala lib
> dep version
> I will beautify it a bit and prepare new version based on it.
>
> I hope this approach will cover your use case nicely.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/prokod/gradle-crossbuild-scala/pull/81#issuecomment-647159113>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABNXZFSGXEYNLFGO4MJFJV3RXZBEXANCNFSM4N47APYQ>
> .
>

borissmidt commented:
So technically it would now make gradle work like sbt? Where it
automatically resolves the _? To the right version for the main source set.

On Sun, Jun 21, 2020, 21:15 boris smidt <smidtboris1@gmail.com> wrote:

> Yes sure thats good as wel! I think its a big boon for this if you have
> apps that define there api in the same project as the app. Since the app
> never has to be cross compiled.
>
> On Sun, Jun 21, 2020, 19:43 Noam Asor <notifications@github.com> wrote:
>
>> @borissmidt <https://github.com/borissmidt> Hi there :)
>>
>> I have been reflecting on the subject and here are some things I
>> resurfaced as reasoning for some of the past decisions
>>
>>    - The name of the plugin was not specifically named ...scala.. as I
>>    have used the plugin in a muti dimensional crosscompiling env. I am working
>>    a lot with Spark and when developing Spark libs quickly you find yourself
>>    maintaining not only multiple scala version but also multiple Spark
>>    versions.
>>    - Related to first topic, this is why your approach as is won't be
>>    fully satisfying, at least from DSL point of view.
>>
>> Last but not least! I have found a different way to tackle that problem,
>> implemented it and ran the tests including the one you added where one of
>> the sub projects does not have crossbuilding plugin applied to it and it
>> passed. No need to add nothing to DSL. It is of course inspired from the
>> need you have expressed and I believe a lot will benefit from !
>> Basically I have added a stage where I auto convert compile lib_? to a
>> set of explicit dependencies that can be resolved by any non crossbuild sub
>> project:
>>
>> crossBuildV211Compile lib_2.11
>> crossBuildV212Compile lib_2.12
>> compile lib_<default_scala_version>
>>
>> where default_scala_version is inferred from the non crossbuild scala lib
>> dep version
>> I will beautify it a bit and prepare new version based on it.
>>
>> I hope this approach will cover your use case nicely.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/prokod/gradle-crossbuild-scala/pull/81#issuecomment-647159113>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ABNXZFSGXEYNLFGO4MJFJV3RXZBEXANCNFSM4N47APYQ>
>> .
>>
>

prokod commented:
@borissmidt It seems so :)

So I am basically done. hopefully later today I will put in maven staging area new version. I would appreciate your feedback

Thanks for the insights and cooperation.

prokod commented:
Hi @borissmidt 
Please checkout new release here https://oss.sonatype.org/content/repositories/comgithubprokod-1036/

borissmidt commented:
How do i include that repository in gradle? i tried
 ```
        maven {
            url "https://oss.sonatype.org/content/repositories/comgithubprokod-1036"
        }
```

prokod commented:
Look at the readme on legacy plugin application:
```
buildscript {
   repositories {
        maven {
            url 'https://oss.sonatype.org/content/repositories/comgithubprokod-1036'
        }
    }
    dependencies {
        classpath("com.github.prokod:gradle-crossbuild-scala:0.11.0")
    }
}
```

borissmidt commented:
I was applying it to the wrong project goodmorning i would say :)

borissmidt commented:
Hey sorry i didn't have much time last week i tried out the new plugin version but i get an error:
```
Caused by: java.util.NoSuchElementException: Cannot access first() element from an empty Iterable
	at com.github.prokod.gradle.crossbuild.CrossBuildPlugin$_globDependencyTranslationForMainSourceSetsConfigurations_closure9$_closure16.doCall(CrossBuildPlugin.groovy:149)
	at com.github.prokod.gradle.crossbuild.CrossBuildPlugin$_globDependencyTranslationForMainSourceSetsConfigurations_closure9.doCall(CrossBuildPlugin.groovy:113)
	at com.github.prokod.gradle.crossbuild.CrossBuildPlugin.globDependencyTranslationForMainSourceSetsConfigurations(CrossBuildPlugin.groovy:102)
	at com.github.prokod.gradle.crossbuild.CrossBuildPlugin$_apply_closure3.doCall(CrossBuildPlugin.groovy:63)
	at org.gradle.configuration.internal.DefaultListenerBuildOperationDecorator$BuildOperationEmittingClosure$1.lambda$run$0(DefaultListenerBuildOperationDecorator.java:180)
	at org.gradle.configuration.internal.DefaultUserCodeApplicationContext.reapply(DefaultUserCodeApplicationContext.java:60)
	at org.gradle.configuration.internal.DefaultListenerBuildOperationDecorator$BuildOperationEmittingClosure$1.run(DefaultListenerBuildOperationDecorator.java:177)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:402)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:394)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:92)
	at org.gradle.configuration.internal.DefaultListenerBuildOperationDecorator$BuildOperationEmittingClosure.doCall(DefaultListenerBuildOperationDecorator.java:174)
	at org.gradle.listener.ClosureBackedMethodInvocationDispatch.dispatch(ClosureBackedMethodInvocationDispatch.java:41)
	at org.gradle.listener.ClosureBackedMethodInvocationDispatch.dispatch(ClosureBackedMethodInvocationDispatch.java:25)
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:42)
	at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:231)
	at org.gradle.internal.event.BroadcastDispatch$SingletonDispatch.dispatch(BroadcastDispatch.java:150)
	at org.gradle.internal.event.AbstractBroadcastDispatch.dispatch(AbstractBroadcastDispatch.java:58)
	at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:325)
	at org.gradle.internal.event.BroadcastDispatch$CompositeDispatch.dispatch(BroadcastDispatch.java:235)
	at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:141)
	at org.gradle.internal.event.ListenerBroadcast.dispatch(ListenerBroadcast.java:37)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy29.afterEvaluate(Unknown Source)
	at org.gradle.configuration.project.LifecycleProjectEvaluator$NotifyAfterEvaluate$1.execute(LifecycleProjectEvaluator.java:186)
	at org.gradle.configuration.project.LifecycleProjectEvaluator$NotifyAfterEvaluate$1.execute(LifecycleProjectEvaluator.java:183)
	at org.gradle.api.internal.project.DefaultProject.stepEvaluationListener(DefaultProject.java:1446)
	at org.gradle.configuration.project.LifecycleProjectEvaluator$NotifyAfterEvaluate.run(LifecycleProjectEvaluator.java:192)
	... 113 more
```

prokod commented:
superseded by #93
